[GTK] Create WebKitWebExtensionContext

ItsJamie9494 opened the pull request:
#### ca3db97b5961c8ec1a7229b1bd8ceb7a6deed2d0
<pre>
[GTK] Create WebKitWebExtensionContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=300631">https://bugs.webkit.org/show_bug.cgi?id=300631</a>

Reviewed by NOBODY (OOPS!).

This creates a new WebKitWebExtensionContext API for WebKitGtk, along with any private APIs in WebKitSettings, WebKitNavigationAction, and WebKitWebView that will be used by WebKitWebExtensionContext. A test suite is provided for the new WebKitWebExtensionContext classes. This also creates a Gtk-specific implementation of WebKit::WebExtension, as well as necessary fixes to WebExtensions code to allow for usage on GTK

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_context_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNavigationAction.cpp:
(webkitNavigationActionGetAction):
* Source/WebKit/UIProcess/API/glib/WebKitNavigationActionPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionContextError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkitSettingsGetWebExtensionContext):
(webkitSettingsSetWebExtensionContext):
* Source/WebKit/UIProcess/API/glib/WebKitSettingsPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
(webkitWebExtensionToImpl):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionContext.cpp: Added.
(_WebKitWebExtensionContextPermission::_WebKitWebExtensionContextPermission):
(webKitWebExtensionContextPermissionCreate):
(webkit_web_extension_context_permission_new):
(webkit_web_extension_context_permission_ref):
(webkit_web_extension_context_permission_unref):
(webkit_web_extension_context_permission_get_permission_name):
(webkit_web_extension_context_permission_get_expiration_date):
(_WebKitWebExtensionContextMatchPattern::_WebKitWebExtensionContextMatchPattern):
(webKitWebExtensionContextMatchPatternCreate):
(webkit_web_extension_context_match_pattern_new):
(webkit_web_extension_context_match_pattern_ref):
(webkit_web_extension_context_match_pattern_unref):
(webkit_web_extension_context_match_pattern_get_match_pattern):
(webkit_web_extension_context_match_pattern_get_expiration_date):
(webkitWebExtensionContextGetProperty):
(webkitWebExtensionContextSetProperty):
(webkit_web_extension_context_class_init):
(webkitWebExtensionContextInitableInit):
(gInitableInterfaceInit):
(webkitWebExtensionContextSetWebExtension):
(webkit_web_extension_context_new_for_extension):
(webkit_web_extension_context_get_web_extension):
(webkit_web_extension_context_get_base_uri):
(webkit_web_extension_context_set_base_uri):
(webkit_web_extension_context_get_unique_identifier):
(webkit_web_extension_context_set_unique_identifier):
(webkit_web_extension_context_get_is_inspectable):
(webkit_web_extension_context_set_is_inspectable):
(webkit_web_extension_context_get_inspection_name):
(webkit_web_extension_context_set_inspection_name):
(webkit_web_extension_context_get_unsupported_apis):
(webkit_web_extension_context_set_unsupported_apis):
(webkit_web_extension_context_get_options_page_uri):
(webkit_web_extension_context_get_override_new_tab_page_uri):
(webkit_web_extension_context_get_granted_permissions):
(webkit_web_extension_context_set_granted_permissions):
(webkit_web_extension_context_get_granted_permission_match_patterns):
(webkit_web_extension_context_set_granted_permission_match_patterns):
(webkit_web_extension_context_get_denied_permissions):
(webkit_web_extension_context_set_denied_permissions):
(webkit_web_extension_context_get_denied_permission_match_patterns):
(webkit_web_extension_context_set_denied_permission_match_patterns):
(webkit_web_extension_context_get_has_requested_optional_access_to_all_hosts):
(webkit_web_extension_context_set_has_requested_optional_access_to_all_hosts):
(webkit_web_extension_context_get_has_access_to_private_data):
(webkit_web_extension_context_set_has_access_to_private_data):
(webkit_web_extension_context_get_current_permissions):
(webkit_web_extension_context_get_current_permission_match_patterns):
(webkit_web_extension_context_has_permission):
(webkit_web_extension_context_has_access_to_uri):
(webkit_web_extension_context_get_has_access_to_all_uris):
(webkit_web_extension_context_get_has_access_to_all_hosts):
(webkit_web_extension_context_get_has_injected_content):
(webkit_web_extension_context_has_injected_content_for_uri):
(webkit_web_extension_context_get_has_content_modification_rules):
(toAPI):
(toImpl):
(webkit_web_extension_context_permission_status_for_permission):
(webkit_web_extension_context_permission_status_for_uri):
(webkit_web_extension_context_permission_status_for_match_pattern):
(webkit_web_extension_context_set_permission_status_for_permission):
(webkit_web_extension_context_set_permission_status_for_uri):
(webkit_web_extension_context_set_permission_status_for_match_pattern):
(webkit_web_extension_context_load_background_content):
(webkit_web_extension_context_load_background_content_finish):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionContext.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
(webkitWebExtensionMatchPatternToImpl):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionPrivate.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewCreatePageConfiguration):
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate _webView:navigationDidFinishDocumentLoad:]):
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
(WebKit::WebExtensionContext::didFinishDocumentLoad):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::isNotRunningInTestRunner): Deleted.
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::plistFileName):
(WebKit::WebExtensionContext::backgroundWebView const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp:
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/glib/WebExtensionContextGLib.cpp: Added.
(decidePolicyCb):
(didFinishDocumentLoadCb):
(didFailNavigationCb):
(webProcessTerminatedCb):
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::recordError):
(WebKit::WebExtensionContext::clearError):
(WebKit::WebExtensionContext::currentState const):
(WebKit::WebExtensionContext::readStateFromPath):
(WebKit::WebExtensionContext::readLastBaseURLFromState):
(WebKit::WebExtensionContext::readDisplayNameFromState):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::writeStateToStorage const):
(WebKit::WebExtensionContext::moveLocalStorageIfNeeded):
(WebKit::WebExtensionContext::setInspectable):
(WebKit::permissionNotification):
(WebKit::WebExtensionContext::permissionsDidChange):
(WebKit::WebExtensionContext::backgroundPageIdentifier const):
(WebKit::WebExtensionContext::relatedWebView):
(WebKit::WebExtensionContext::webViewConfiguration):
(WebKit::WebExtensionContext::isBackgroundPage const):
(WebKit::WebExtensionContext::backgroundContentIsLoaded const):
(WebKit::WebExtensionContext::loadBackgroundWebViewIfNeeded):
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::setBackgroundWebViewInspectionName):
(WebKit::WebExtensionContext::unloadBackgroundContentIfPossible):
(WebKit::WebExtensionContext::unloadBackgroundWebView):
(WebKit::WebExtensionContext::determineInstallReasonDuringLoad):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::decidePolicyForNavigationAction):
(WebKit::WebExtensionContext::didFinishDocumentLoad):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
(WebKit::WebExtensionContext::processes const):
(WebKit::WebExtensionContext::declarativeNetRequestContentRuleListFilePath):
(WebKit::WebExtensionContext::setSessionStorageAllowedInContentScripts):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/Tests/WebKit/WKPage/glib/TestWebKitWebExtensionContext.cpp: Added.
(createGBytes):
(testDefaultPermissionChecks):
(testPermissionGranting):
(testContentScriptsParsing):
(testOptionsPageURIParsing):
(testURIOverridesParsing):
(testLoadBackgroundContent):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca3db97b5961c8ec1a7229b1bd8ceb7a6deed2d0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/174631 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/48564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/42345 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/184209 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/132499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/176496 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/49856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/48247 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/135873 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/95883 "1 flakes 3 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/177589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/39304 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/156664 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/116351 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/173952 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/37945 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/36324 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/32689 "Built successfully") | 
| [✅ 🧪 jsc-x86-64](https://ews-build.webkit.org/#/builders/218/builds/5173 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/148368 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/36156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/186636 "Built successfully") | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/35893 "Built successfully and passed tests") | [❌ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/39929 "Found 1028 new failures in b3/testb3_7.cpp, b3/B3Value.h, WebProcess/Network/WebResourceLoader.cpp, jit/OperationResult.h, UIProcess/WebProcessPool.cpp, Shared/Cocoa/WKNSArray.mm, heap/WeakSet.h, API/tests/Regress141275.mm, Shared/SessionState.cpp, NetworkProcess/NetworkResourceLoadParameters.cpp ...") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/40522 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/144798 "Found 1 new test failure: ipc/async-with-reply-destination-id-zero.html (failure)") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/48231 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/144657 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/48910 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/32777 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/114690 "The change is no longer eligible for processing.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/39530 "Passed tests") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/120916 "Found 121 new failures in b3/B3Value.h, b3/testb3_7.cpp, WebProcess/Network/WebResourceLoader.cpp, UIProcess/WebProcessPool.cpp, Shared/Cocoa/WKNSArray.mm, Shared/SessionState.cpp, NetworkProcess/NetworkResourceLoadParameters.cpp, WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.mm, WebProcess/WebPage/WebPage.cpp, WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm ...") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/211684 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/47417 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/11124 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/55223 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/46819 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/47050 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/46877 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/1fa4dd6e014a70716f8381bd674f276043a836f4)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126373 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128244 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129321 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/125725 "Passed tests") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

ItsJamie9494 commented on Source/WebKit/UIProcess/API/glib/WebKitSettingsPrivate.h:
```diff
@@ -30,9 +30,25 @@
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+WK_EXPORT RefPtr<WebKit::WebExtensionContext> webkitSettingsGetWebExtensionContext(WebKitSettings*);
+WK_EXPORT void webkitSettingsSetWebExtensionContext(WebKitSettings*, RefPtr<WebKit::WebExtensionContext>);
```
const Foo& would probably be best here, the settings should just be holding a weak reference to the Context to allow for some functions to be called from inside WebKitWebView

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -845,6 +845,21 @@ static Ref<API::PageConfiguration> webkitWebViewCreatePageConfiguration(WebKitWe
+        pageConfiguration->setCrossOriginAccessControlCheckEnabled(false);
+        pageConfiguration->setProcessDisplayName(context->processDisplayName());
+        pageConfiguration->setRequiredWebExtensionBaseURL(URL(context->baseURL()));
+        pageConfiguration->setShouldRelaxThirdPartyCookieBlocking(WebCore::ShouldRelaxThirdPartyCookieBlocking::Yes);
```
Since this is disabling security features this deserves a comment.

I'm not sure that these are long term solutions right?

ItsJamie9494 replied:
I agree, but I'd rather keep parity with Cocoa. We can investigate whether this is required on both platforms later

TingPing replied:
Extension pages still have limited cross-origin access based on their permissions. I haven't looked where WebKit handles that, so I'm not saying there is an issue. I mostly wanted a comment to clarify why this cross-origin property was disabled to know it was considered. The Cocoa port doing so eases that worry, but still for future readers it's nice to know in case things get out of sync.

ItsJamie9494 commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionContextGTK.cpp:
```diff
@@ -0,0 +1,160 @@
+    bool isManifestVersion3 = protectedExtension()->supportsManifestVersion(3);
+
+    WebKitSettings* settings = webViewConfiguration();
+    WebKitWebView* webView = WEBKIT_WEB_VIEW(g_object_new(WEBKIT_TYPE_WEB_VIEW,
```
Both `WebContext` and `NetworkSession` are already available as construct-only props? (https://webkitgtk.org/reference/webkitgtk/stable/property.WebView.web-context.html, https://webkitgtk.org/reference/webkitgtk/stable/property.WebView.network-session.html). Do you mean WebExtensionContext?

TingPing replied:
I think `WebExtensionContext` needs access to a `WebContext`, as the browser may be using a custom one, and this choosing the default context every time a view is created may be incorrect.

ItsJamie9494 replied:
Yes, so reasonably just NetworkSession and WebContext should be created here and passed into the WebView. I know we use DataManager at some point.

ItsJamie9494 replied:
Ah, right, so we'd need to expose construct-only props on the `WebExtensionContext` for this

xeenon replied:
In Cocoa, these things are controlled by the `WKWebExtensionControllerConfiguration` and this property:

```objc
/*! @abstract The web view configuration to be used as a basis for configuring web views in extension contexts. */
@property (nonatomic, null_resettable, copy) WKWebViewConfiguration *webViewConfiguration;
```

I don't think it makes sense to be configured per `WebExtensionContext`.

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/1f96a279617d57c3cf187e15d646635b1e28bed4)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132002 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4494 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139516 "Build is in progress. Recent messages:") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133872 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4437 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4256 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/139516 "Build is in progress. Recent messages:") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134948 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/4437 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/139516 "Build is in progress. Recent messages:") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/131351 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/4437 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82734 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/4437 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142161 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/142161 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4245 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/4256 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/142161 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/43018 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-26-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4163,6 +4166,59 @@ void webkitSettingsSetGetUserMediaRequiresFocus(WebKitSettings* settings, bool r
+    if (priv->webExtensionContext)
+        return priv->webExtensionContext.get();
+    return nullptr;
+}
```
Why do we save this in settings?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettingsPrivate.h:
```diff
@@ -30,9 +30,25 @@
+WK_EXPORT void webkitSettingsSetPageVisibilityBasedProcessSuppressionEnabled(WebKitSettings*, bool enabled);
+WK_EXPORT void webkitSettingsSetShouldTakeNearSuspendedAssertions(WebKitSettings*, bool enabled);
+WK_EXPORT void webkitSettingsSetBackgroundWebContentRunningBoardThrottlingEnabled(WebKitSettings*, bool enabled);
+WK_EXPORT void webkitSettingsSetShouldDropNearSuspendedAssertionAfterDelay(WebKitSettings*, bool enabled);
```
I don't think we want to add one wrapper for every internal setting when we already have GetPreferences

ItsJamie9494 committed:
[GTK] Create WebKitWebExtensionContext
https://bugs.webkit.org/show_bug.cgi?id=300631

Reviewed by NOBODY (OOPS!).

This creates a new WebKitWebExtensionContext API for WebKitGtk, along with any private APIs in WebKitSettings, WebKitNavigationAction, and WebKitWebView that will be used by WebKitWebExtensionContext. A test suite is provided for the new WebKitWebExtensionContext classes. This also creates a Gtk-specific implementation of WebKit::WebExtension, as well as necessary fixes to WebExtensions code to allow for usage on GTK

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_context_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNavigationAction.cpp:
(webkitNavigationActionGetAction):
* Source/WebKit/UIProcess/API/glib/WebKitNavigationActionPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionContextError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkitSettingsGetSiteIsolationEnabled):
(webkitSettingsSetHiddenPageDOMTimerThrottlingEnabled):
(webkitSettingsSetPageVisibilityBasedProcessSuppressionEnabled):
(webkitSettingsSetShouldTakeNearSuspendedAssertions):
(webkitSettingsSetBackgroundWebContentRunningBoardThrottlingEnabled):
(webkitSettingsSetShouldDropNearSuspendedAssertionAfterDelay):
(webkitSettingsGetWebExtensionContext):
(webkitSettingsSetWebExtensionContext):
* Source/WebKit/UIProcess/API/glib/WebKitSettingsPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
(webkitWebExtensionToImpl):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionContext.cpp: Added.
(_WebKitWebExtensionContextPermission::_WebKitWebExtensionContextPermission):
(webKitWebExtensionContextPermissionCreate):
(webkit_web_extension_context_permission_new):
(webkit_web_extension_context_permission_ref):
(webkit_web_extension_context_permission_unref):
(webkit_web_extension_context_permission_get_permission_name):
(webkit_web_extension_context_permission_get_expiration_date):
(_WebKitWebExtensionContextMatchPattern::_WebKitWebExtensionContextMatchPattern):
(webKitWebExtensionContextMatchPatternCreate):
(webkit_web_extension_context_match_pattern_new):
(webkit_web_extension_context_match_pattern_ref):
(webkit_web_extension_context_match_pattern_unref):
(webkit_web_extension_context_match_pattern_get_match_pattern):
(webkit_web_extension_context_match_pattern_get_expiration_date):
(webkitWebExtensionContextGetProperty):
(webkitWebExtensionContextSetProperty):
(webkit_web_extension_context_class_init):
(webkitWebExtensionContextInitableInit):
(gInitableInterfaceInit):
(webkitWebExtensionContextSetWebExtension):
(webkit_web_extension_context_get_backend):
(webkit_web_extension_context_get_web_extension):
(webkit_web_extension_context_get_base_uri):
(webkit_web_extension_context_set_base_uri):
(webkit_web_extension_context_get_unique_identifier):
(webkit_web_extension_context_set_unique_identifier):
(webkit_web_extension_context_get_is_inspectable):
(webkit_web_extension_context_set_is_inspectable):
(webkit_web_extension_context_get_inspection_name):
(webkit_web_extension_context_set_inspection_name):
(webkit_web_extension_context_get_unsupported_apis):
(webkit_web_extension_context_set_unsupported_apis):
(webkit_web_extension_context_get_options_page_uri):
(webkit_web_extension_context_get_override_new_tab_page_uri):
(webkit_web_extension_context_get_granted_permissions):
(webkit_web_extension_context_set_granted_permissions):
(webkit_web_extension_context_get_granted_permission_match_patterns):
(webkit_web_extension_context_set_granted_permission_match_patterns):
(webkit_web_extension_context_get_denied_permissions):
(webkit_web_extension_context_set_denied_permissions):
(webkit_web_extension_context_get_denied_permission_match_patterns):
(webkit_web_extension_context_set_denied_permission_match_patterns):
(webkit_web_extension_context_get_has_requested_optional_access_to_all_hosts):
(webkit_web_extension_context_set_has_requested_optional_access_to_all_hosts):
(webkit_web_extension_context_get_has_access_to_private_data):
(webkit_web_extension_context_set_has_access_to_private_data):
(webkit_web_extension_context_get_current_permissions):
(webkit_web_extension_context_get_current_permission_match_patterns):
(webkit_web_extension_context_has_permission):
(webkit_web_extension_context_has_access_to_uri):
(webkit_web_extension_context_get_has_access_to_all_uris):
(webkit_web_extension_context_get_has_access_to_all_hosts):
(webkit_web_extension_context_get_has_injected_content):
(webkit_web_extension_context_has_injected_content_for_uri):
(webkit_web_extension_context_get_has_content_modification_rules):
(toAPI):
(toImpl):
(webkit_web_extension_context_permission_status_for_permission):
(webkit_web_extension_context_permission_status_for_uri):
(webkit_web_extension_context_permission_status_for_match_pattern):
(webkit_web_extension_context_set_permission_status_for_permission):
(webkit_web_extension_context_set_permission_status_for_uri):
(webkit_web_extension_context_set_permission_status_for_match_pattern):
(webkit_web_extension_context_load_background_content):
(webkit_web_extension_context_load_background_content_finish):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionContext.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
(webkitWebExtensionMatchPatternToImpl):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionPrivate.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewCreatePageConfiguration):
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitWebExtensionContextGtk.cpp: Added.
(webkit_web_extension_context_new_for_extension):
* Source/WebKit/UIProcess/API/wpe/WebKitWebExtensionContextWPE.cpp: Added.
(webkit_web_extension_context_new_for_extension):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate _webView:navigationDidFinishDocumentLoad:]):
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
(WebKit::WebExtensionContext::didFinishDocumentLoad):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::isNotRunningInTestRunner): Deleted.
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::plistFileName):
(WebKit::WebExtensionContext::backgroundWebView const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp:
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/glib/WebExtensionContextGLib.cpp: Added.
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::recordError):
(WebKit::WebExtensionContext::clearError):
(WebKit::WebExtensionContext::currentState const):
(WebKit::WebExtensionContext::readStateFromPath):
(WebKit::WebExtensionContext::readLastBaseURLFromState):
(WebKit::WebExtensionContext::readDisplayNameFromState):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::writeStateToStorage const):
(WebKit::WebExtensionContext::moveLocalStorageIfNeeded):
(WebKit::WebExtensionContext::setInspectable):
(WebKit::permissionNotification):
(WebKit::WebExtensionContext::permissionsDidChange):
(WebKit::WebExtensionContext::backgroundPageIdentifier const):
(WebKit::WebExtensionContext::webViewConfiguration):
(WebKit::WebExtensionContext::isBackgroundPage const):
(WebKit::WebExtensionContext::backgroundContentIsLoaded const):
(WebKit::WebExtensionContext::loadBackgroundWebViewIfNeeded):
(WebKit::WebExtensionContext::unloadBackgroundWebView):
(WebKit::WebExtensionContext::setBackgroundWebViewInspectionName):
(WebKit::WebExtensionContext::unloadBackgroundContentIfPossible):
(WebKit::WebExtensionContext::determineInstallReasonDuringLoad):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::decidePolicyForNavigationAction):
(WebKit::WebExtensionContext::didFinishDocumentLoad):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
(WebKit::WebExtensionContext::processes const):
(WebKit::WebExtensionContext::declarativeNetRequestContentRuleListFilePath):
(WebKit::WebExtensionContext::setSessionStorageAllowedInContentScripts):
* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionContextGTK.cpp: Added.
(decidePolicyCb):
(didFinishDocumentLoadCb):
(didFailNavigationCb):
(webProcessTerminatedCb):
(WebKit::WebExtensionContext::relatedWebView):
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
* Source/WebKit/UIProcess/Extensions/wpe/WebExtensionContextWPE.cpp: Added.
(decidePolicyCb):
(didFinishDocumentLoadCb):
(didFailNavigationCb):
(webProcessTerminatedCb):
(WebKit::WebExtensionContext::relatedWebView):
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtensionContext.cpp: Added.
(createGBytes):
(createContext):
(testDefaultPermissionChecks):
(testPermissionGranting):
(testContentScriptsParsing):
(testOptionsPageURIParsing):
(testURIOverridesParsing):
(testLoadBackgroundContent):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:

carlosgcampos reviewed (commented):
I've added a few comments, but it's quite difficult for me to review this huge patch, can't this be split?

ItsJamie9494 committed:
[GTK] Create WebKitWebExtensionContext
https://bugs.webkit.org/show_bug.cgi?id=300631

Reviewed by NOBODY (OOPS!).

This creates a new WebKitWebExtensionContext API for WebKitGtk, along with any private APIs in WebKitSettings, WebKitNavigationAction, and WebKitWebView that will be used by WebKitWebExtensionContext. A test suite is provided for the new WebKitWebExtensionContext classes. This also creates a Gtk-specific implementation of WebKit::WebExtension, as well as necessary fixes to WebExtensions code to allow for usage on GTK

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.h:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_context_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNavigationAction.cpp:
(webkitNavigationActionGetAction):
* Source/WebKit/UIProcess/API/glib/WebKitNavigationActionPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionContextError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkitSettingsGetWebExtensionContext):
(webkitSettingsSetWebExtensionContext):
* Source/WebKit/UIProcess/API/glib/WebKitSettingsPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp:
(webkitWebExtensionToImpl):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionContext.cpp: Added.
(_WebKitWebExtensionContextPermission::_WebKitWebExtensionContextPermission):
(webKitWebExtensionContextPermissionCreate):
(webkit_web_extension_context_permission_new):
(webkit_web_extension_context_permission_ref):
(webkit_web_extension_context_permission_unref):
(webkit_web_extension_context_permission_get_permission_name):
(webkit_web_extension_context_permission_get_expiration_date):
(_WebKitWebExtensionContextMatchPattern::_WebKitWebExtensionContextMatchPattern):
(webKitWebExtensionContextMatchPatternCreate):
(webkit_web_extension_context_match_pattern_new):
(webkit_web_extension_context_match_pattern_ref):
(webkit_web_extension_context_match_pattern_unref):
(webkit_web_extension_context_match_pattern_get_match_pattern):
(webkit_web_extension_context_match_pattern_get_expiration_date):
(webkitWebExtensionContextGetProperty):
(webkitWebExtensionContextSetProperty):
(webkit_web_extension_context_class_init):
(webkitWebExtensionContextInitableInit):
(gInitableInterfaceInit):
(webkitWebExtensionContextSetWebExtension):
(webkit_web_extension_context_new_for_extension):
(webkit_web_extension_context_get_web_extension):
(webkit_web_extension_context_get_base_uri):
(webkit_web_extension_context_set_base_uri):
(webkit_web_extension_context_get_unique_identifier):
(webkit_web_extension_context_set_unique_identifier):
(webkit_web_extension_context_get_is_inspectable):
(webkit_web_extension_context_set_is_inspectable):
(webkit_web_extension_context_get_inspection_name):
(webkit_web_extension_context_set_inspection_name):
(webkit_web_extension_context_get_unsupported_apis):
(webkit_web_extension_context_set_unsupported_apis):
(webkit_web_extension_context_get_options_page_uri):
(webkit_web_extension_context_get_override_new_tab_page_uri):
(webkit_web_extension_context_get_granted_permissions):
(webkit_web_extension_context_set_granted_permissions):
(webkit_web_extension_context_get_granted_permission_match_patterns):
(webkit_web_extension_context_set_granted_permission_match_patterns):
(webkit_web_extension_context_get_denied_permissions):
(webkit_web_extension_context_set_denied_permissions):
(webkit_web_extension_context_get_denied_permission_match_patterns):
(webkit_web_extension_context_set_denied_permission_match_patterns):
(webkit_web_extension_context_get_has_requested_optional_access_to_all_hosts):
(webkit_web_extension_context_set_has_requested_optional_access_to_all_hosts):
(webkit_web_extension_context_get_has_access_to_private_data):
(webkit_web_extension_context_set_has_access_to_private_data):
(webkit_web_extension_context_get_current_permissions):
(webkit_web_extension_context_get_current_permission_match_patterns):
(webkit_web_extension_context_has_permission):
(webkit_web_extension_context_has_access_to_uri):
(webkit_web_extension_context_get_has_access_to_all_uris):
(webkit_web_extension_context_get_has_access_to_all_hosts):
(webkit_web_extension_context_get_has_injected_content):
(webkit_web_extension_context_has_injected_content_for_uri):
(webkit_web_extension_context_get_has_content_modification_rules):
(toAPI):
(toImpl):
(webkit_web_extension_context_permission_status_for_permission):
(webkit_web_extension_context_permission_status_for_uri):
(webkit_web_extension_context_permission_status_for_match_pattern):
(webkit_web_extension_context_set_permission_status_for_permission):
(webkit_web_extension_context_set_permission_status_for_uri):
(webkit_web_extension_context_set_permission_status_for_match_pattern):
(webkit_web_extension_context_load_background_content):
(webkit_web_extension_context_load_background_content_finish):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionContext.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
(webkitWebExtensionMatchPatternToImpl):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionPrivate.h: Copied from Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewCreatePageConfiguration):
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate _webView:navigationDidFinishDocumentLoad:]):
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
(WebKit::WebExtensionContext::didFinishDocumentLoad):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::isNotRunningInTestRunner): Deleted.
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::plistFileName):
(WebKit::WebExtensionContext::backgroundWebView const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp:
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/glib/WebExtensionContextGLib.cpp: Added.
(decidePolicyCb):
(didFinishDocumentLoadCb):
(didFailNavigationCb):
(webProcessTerminatedCb):
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::recordError):
(WebKit::WebExtensionContext::clearError):
(WebKit::WebExtensionContext::currentState const):
(WebKit::WebExtensionContext::readStateFromPath):
(WebKit::WebExtensionContext::readLastBaseURLFromState):
(WebKit::WebExtensionContext::readDisplayNameFromState):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::writeStateToStorage const):
(WebKit::WebExtensionContext::moveLocalStorageIfNeeded):
(WebKit::WebExtensionContext::setInspectable):
(WebKit::permissionNotification):
(WebKit::WebExtensionContext::permissionsDidChange):
(WebKit::WebExtensionContext::backgroundPageIdentifier const):
(WebKit::WebExtensionContext::relatedWebView):
(WebKit::WebExtensionContext::webViewConfiguration):
(WebKit::WebExtensionContext::isBackgroundPage const):
(WebKit::WebExtensionContext::backgroundContentIsLoaded const):
(WebKit::WebExtensionContext::loadBackgroundWebViewIfNeeded):
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::setBackgroundWebViewInspectionName):
(WebKit::WebExtensionContext::unloadBackgroundContentIfPossible):
(WebKit::WebExtensionContext::unloadBackgroundWebView):
(WebKit::WebExtensionContext::determineInstallReasonDuringLoad):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::decidePolicyForNavigationAction):
(WebKit::WebExtensionContext::didFinishDocumentLoad):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
(WebKit::WebExtensionContext::processes const):
(WebKit::WebExtensionContext::declarativeNetRequestContentRuleListFilePath):
(WebKit::WebExtensionContext::setSessionStorageAllowedInContentScripts):
(WebKit::WebExtensionContext::isNotRunningInTestRunner):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/Tests/WebKit/WKPage/glib/TestWebKitWebExtensionContext.cpp: Added.
(createGBytes):
(testDefaultPermissionChecks):
(testPermissionGranting):
(testContentScriptsParsing):
(testOptionsPageURIParsing):
(testURIOverridesParsing):
(testLoadBackgroundContent):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
